File: Linked list/Reorder List/Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
      list<ListNode*> mylist;
      ListNode* temp = head;
      while( temp != NULL ){
         mylist.push_back(temp);
         temp = temp->next;
      }
     if(mylist.size()>1){
      ListNode* finl = head;
      ListNode* res = finl;
      ListNode* front = mylist.front();
      ListNode* back = mylist.back();
      finl = front;
      finl->next = back;
      finl = finl->next;
      mylist.pop_front();
      mylist.pop_back();
      while( mylist.size() > 0 && finl != NULL){
          front = mylist.front();
          back = mylist.back();
          finl->next = front;
          finl = finl->next;
          finl->next = back;          
          finl = finl->next;
          if(mylist.size()>0)
           mylist.pop_front();
          
          if(mylist.size()>0)
           mylist.pop_back();
       }
        finl->next = NULL;
        head = res;
     }
    }
};

File: Linked list/Reorder List/Reorder List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ArrayList<ListNode> mylist = new ArrayList<ListNode>(); 
        ListNode temp = head;
        while( temp != null ){
         mylist.add(temp);
         temp = temp.next;
        }
        if( mylist.size() > 1){
          ListNode finl = head;
          ListNode res = finl;
          ListNode front = mylist.get(0);
          ListNode back = mylist.get( mylist.size() -1 );
          finl = front;
          finl.next = back;
          finl = finl.next;
          mylist.remove(0);
          mylist.remove( mylist.size() - 1 );
             while( mylist.size() > 0 && finl != null ){
              front = mylist.get(0);
              back = mylist.get( mylist.size() - 1 );
              finl.next = front;
              finl = finl.next;
              finl.next = back;          
              finl = finl.next;
              if(mylist.size()>0)
               mylist.remove(0);
              if(mylist.size()>0)
               mylist.remove( mylist.size() - 1 );
           }
            finl.next = null;
            head = res;
        }
    }
}
